### ISSUE TITLE
[$500] Workspace - Endless skeleton loader on the member page if you switch to it multiple times

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#40595]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.63.7
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Log in with a new expensifail account
2. Create a workspace
3. Navigate to the members page
4. Navigate back
5. Navigate to the members page again and wait 20-30 seconds
6. Navigate back
7. Navigate to the members page again

## Expected Result:
Skeleton loader should only be visible for a very short time

## Actual Result:
Endless skeleton loader appears on the member page if you switch to it multiple times

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/fe94dd0a-995c-4b64-ade9-cc70c763e0b2

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e843a4238c37c0d9</li>
        <li>Upwork Job ID: 1781491296362684416</li>
        <li>Last Price Increase: 2024-05-15</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
@puneetlath  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by mericanAncer127** [comment]
Upon thorough analysis, it's evident that the endless skeleton loader arises from delays in data fetching and rendering when switching to the member page multiple times. To address this issue, I propose the following technical changes:

1. I'll implement efficient data fetching strategies aimed at reducing latency and improving loading times. This may involve optimizing network requests, employing caching mechanisms, or exploring asynchronous data loading techniques.
2. By revisiting the component rendering logic, I'll prioritize the display of essential content while asynchronously loading additional data in the background. This approach will mitigate the appearance of the endless skeleton loader and provide users with meaningful content promptly.
3. I'll introduce robust error handling mechanisms to gracefully manage data retrieval failures and network interruptions. Users will be presented with informative error messages, and detailed error logs will be generated for debugging purposes.
4. Integration of performance monitoring tools will be crucial for tracking page load times and identifying potential performance bottlenecks. Continuous monitoring will enable me to refine the member page loading experience and maintain optimal performance levels.

I hope this solution will help you with this issue.

Cooper Brown from Upwork.

**Comment by jitendraCredentia** [comment]
**Issue Analysis and Proposal**:

After reviewing the reported problem and examining the behavior described, it appears that the infinite scroll feature may be closely related to the issue at hand. The endless skeleton loader observed on the member page could potentially be triggered or exacerbated by the continuous loading mechanism implemented through infinite scroll.

**Understanding the Impact:**

Infinite scroll, while convenient for users in certain contexts, can sometimes lead to performance issues or unexpected behaviors, especially when dealing with complex data rendering and asynchronous loading processes. In this case, the endless skeleton loader suggests a failure in properly handling the continuous loading of member data.

**Proposal for Resolution:**

To address this issue effectively, I propose temporarily disabling the infinite scroll feature on the member page. By doing so, we can isolate the behavior and investigate any underlying causes more thoroughly. Additionally, this approach will allow us to assess the impact of infinite scroll on the reported problem and determine whether it is the primary contributing factor.

**Next Steps:**

Disable the infinite scroll feature on the member page.
Conduct comprehensive testing to observe any changes in behavior and performance.
Investigate the root cause of the endless skeleton loader issue in parallel.
Implement necessary fixes or optimizations based on the findings.
Re-enable infinite scroll once the underlying issues are resolved and thoroughly tested.

**Comment by mountiny** [comment]
This feels like a niche issue so if by the time other blockers are resolved, I would demote it

**Comment by hungvu193** [comment]
I can't reproduce on latest main

**Comment by mountiny** [comment]
I will demote now and add Needs reproduction label

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by puneetlath** [comment]
I've asked applause to re-test.

**Comment by kavimuru** [comment]
Tester is still reproducible in step 5. Also it is reproducible in web. 

https://github.com/Expensify/App/assets/43996225/62db99e1-01ae-4dac-afdc-6b53a2bd12d8

**Comment by puneetlath** [comment]
Ok, let's keep waiting on proposals then.

**Comment by puneetlath** [comment]
Bumping this to $500 to try and get a proposal.

**Comment by dominictb** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Endless skeleton loader appears on the member page if you switch to it multiple times

### What is the root cause of that problem?

I believe that there're only 1 scenario that lead to this issue, which is somehow API calls `CreateWorkspace` and `OpenWorkspaceMembersPage` return `{jsonCode !== 200}` and the `getMissingOnyxUpdates` call fails to pick up the missing updates from the server. I can partially simulate this locally by setting in the proxy:

```typescript
   proxyRequest.on('response', (proxyResponse) => {
      if(requestPath?.includes('CreateWorkspace') || requestPath?.includes('OpenWorkspaceMembersPage')){
          response.writeHead(200, {'content-type': 'application/json'});
          response.write(Buffer.from(JSON.stringify({jsonCode: 429})))
          response.end()
          return;
      }
      response.writeHead(proxyResponse.statusCode ?? 0, proxyResponse.headers);
      proxyResponse.pipe(response);
  });
```
If the `CreateWorkspace` fails, it will apply the onyx update
```
const failureData: OnyxUpdate[] = [
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
            value: {employeeList: null},
        },
...
```
causing the `policy.employeeList` to be null, and since `OpenWorkspaceMembersPage` fails too, the `employeeList` is still `null`, causing `isLoading` value in the component `WorkspaceMembersPage` to be `false`. Below is the definition of `isLoading`

```typescript
const isOfflineAndNoMemberDataAvailable = isEmptyObject(policy?.employeeList) && isOffline;

const isLoading = useMemo(
        () => !isOfflineAndNoMemberDataAvailable && (!OptionsListUtils.isPersonalDetailsReady(personalDetails) || isEmptyObject(policy?.employeeList)),
        [isOfflineAndNoMemberDataAvailable, personalDetails, policy?.employeeList],
    );
```

We can guarantee that `OptionsListUtils.isPersonalDetailsReady(personalDetails)` should always be true, and with `isOffline = false` and `policy?.employeeList = null`, the `isLoading` is always true in this case => the skeleton loader will be rendered in the `SelectionList` component

```typescript
<SelectionList
                        ref={selectionListRef}
                        canSelectMultiple={isPolicyAdmin}
                        sections={[{data, isDisabled: false}]}
                        ListItem={TableListItem}
                        disableKeyboardShortcuts={removeMembersConfirmModalVisible}
                        headerMessage={getHeaderMessage()}
                        headerContent={getHeaderContent()}
                        onSelectRow={openMemberDetails}
                        onCheckboxPress={(item) => toggleUser(item.accountID)}
                        onSelectAll={() => toggleAllUsers(data)}
                        onDismissError={dismissError}
                        showLoadingPlaceholder={isLoading}
                        showScrollIndicator
                        shouldPreventDefaultFocusOnSelectRow={!DeviceCapabilities.canUseTouchScreen()}
                        textInputRef={textInputRef}
                        customListHeader={getCustomListHeader()}
                        listHeaderWrapperStyle={[styles.ph9, styles.pv3, styles.pb5]}
                    />
```

You can see the video below to see the described behavior (until the `getMissingOnyxUpdate` call finish, which is the call I haven't been able to simulate yet)

https://github.com/Expensify/App/assets/165644294/9e48c831-759a-4d77-9296-d25d2be95b3d


### What changes do you think we should make in order to solve the problem?


First, there are no concrete evidence supporting my assumption about the failures of these 2 APIs in this bug (and the `GetMissingOnyxUpdate` failure to pick up missing update sequence as well), even though via elimination process, I can safely conclude that only under this circumstances that we can reproduce this bug.

<details>
  <summary><i>Reasoning</i></summary>

We'll need to take a look at the `isLoading` definition
```typescript
function isPersonalDetailsReady(personalDetails: OnyxEntry<PersonalDetailsList>): boolean {
    const personalDetailsKeys = Object.keys(personalDetails ?? {});
    return personalDetailsKeys.some((key) => personalDetails?.[key]?.accountID);
}

const isOfflineAndNoMemberDataAvailable = isEmptyObject(policy?.employeeList) && isOffline;

const isLoading = useMemo(
    () => !isOfflineAndNoMemberDataAvailable && (!OptionsListUtils.isPersonalDetailsReady(personalDetails) || isEmptyObject(policy?.employeeList)),
    [isOfflineAndNoMemberDataAvailable, personalDetails, policy?.employeeList],
);
```

The definition of the `isPersonalDetailsReady` shows that it is impossible for `isPersonalDetailsReady(personalDetails)` to be false, in order for `isLoading=true`, the `isOffline` must be false and `isEmptyObject(policy?.employeeList)` must be empty. By looking through all possible Onyx updates, we can assume that only when the `CreateWorkspace` call fails [function createWorkspace](https://github.com/Expensify/App/blob/550e83b2f55f0998feece27b1768ad7148d03398/src/libs/actions/Policy.ts#L2370) and the `failureData` are applied, then the `policy?.employeeList` can be null. 

Further investigation into the optimistic update of CreateWorkspace API call, and other related API calls like `OpenWorkspaceMembersPage` shows the exact behavior as described in the all of the reported videos [here](https://github.com/Expensify/App/issues/40595#issuecomment-2079194747) : At first opening the member page (and backing out many times), we can still see the member list (due to the optimistic data), but after roughly dozens of seconds (roughly the `CreateWorkspace` API call response time, the `OpenWorkspaceMembersPage` seems to be faster, but it could be slow in the reported video), it shows skeleton loader (onyx pick up `failureData` updates, causing the `employeeList` to be null). 

</details>

If the assumption is true, the BE should take a look into the root cause of API failures, but in the FE app it is safe to say the `isLoading` could always be `false`. Even in the case of API failures, we can still display some message like `fail to load the member list` when `employeeList = null`, which is plausible since every workspace should have at least 1 member (which is the owner). Hence, we can remove the `isLoading` in the `WorkspaceMembersPage` component

### What alternative solutions did you explore? (Optional)

N/A

**Comment by dominictb** [comment]
https://github.com/Expensify/App/issues/40595#issuecomment-2094332042 man if this is true, I swear this bug worth `$1000` bounty

**Comment by hungvu193** [comment]
@dominictb Do you have any consistent steps to reproduce this issue?

**Comment by dominictb** [comment]
@hungvu193 Currently no, as mentioned, I could only simulate (but not reproduce) the issue based on my assumption.

**Comment by hungvu193** [comment]
tbh, that's really hard to review it while we couldn't reproduce it but let me try to take another look today.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by dominictb** [comment]
@hungvu193 u got any chance to take a look at this?

**Comment by hungvu193** [comment]
> and since the workspace member list cannot be empty (at least the owner is in the member list), we can show the message `Cannot display the member list right now! Please try again!` (this is subject to change) by modifying the `getHeaderMessage` function
> 
> 
> ```
>  const getHeaderMessage = () => {
>     if (isOfflineAndNoMemberDataAvailable) {
>         return translate('workspace.common.mustBeOnlineToViewMembers');
>     }
> 
>     return isEmptyObject(policy?.employeeList) ? 'Cannot display the member list right now! Please try again!': '';
> };
> ```

Still can not reproduce this issue but I don't agree with you at this point, `employeeList` can be empty during it's fetching from server for the first time user open the app.

<img width="492" alt="Screenshot 2024-05-08 at 14 50 46" src="https://github.com/Expensify/App/assets/16502320/1cafd233-486e-400d-958b-3d9fcdb2bc37">

Since this issue was not reproduced anymore, I just wonder if we can close it?

**Comment by dominictb** [comment]
@hungvu193 the empty array value of `employeeList` comes from the OpenApp API, the `mergeCollection` for the `policy_` collection operation. It's a separate issue and the BE should take a look into this?

<img width="526" alt="image" src="https://github.com/Expensify/App/assets/165644294/8af08966-22e7-461f-9dc9-06b885f2fd5a">

However, if the OpenApp returns, the employeeList again cannot be empty, or at least it cannot be null unless the situation described in my proposal happens, thus my assumption is not totally inaccurate

**Comment by hungvu193** [comment]
> the empty array value of `employeeList` comes from the OpenApp API,

I think this is intentional. So I still don't agree with you at that point.
cc @puneetlath for your thoughts at above discussion, I think we can close this issue.

**Comment by puneetlath** [comment]
Ok, based on the discussion I agree that we can close this out. Thanks everyone.
